packages/core/utils/src/dml/__tests__/base_relationship.spec.ts

```diff
@@ -21,6 +21,37 @@ describe("Base relationship", () => {
     expect(relationship.parse("user")).toEqual({
       name: "user",
       type: "hasOne",
+      nullable: false,
+      optional: false,
+      entity: entityRef,
+      options: {
+        foreignKey: "user_id",
+      },
+    })
+  })
+
+  test("mark relationship as nullable", () => {
+    class HasOne<T> extends BaseRelationship<T> {
+      protected relationshipType: "hasOne" | "hasMany" | "manyToMany" = "hasOne"
+    }
+
+    const user = {
+      username: new TextSchema(),
+    }
+
+    const entityRef = () => user
+    const relationship = new HasOne(entityRef, {
+      foreignKey: "user_id",
+    }).nullable()
+
+    expectTypeOf(relationship["$dataType"]).toEqualTypeOf<
+      (() => typeof user) | null
+    >()
+    expect(relationship.parse("user")).toEqual({
+      name: "user",
+      type: "hasOne",
+      nullable: true,
+      optional: false,
```

Comment:
Q: if it is nullable, shouldn't it be optional? in case you create a relation, you shouldn't be forced to pass the user

Reply:
Ideally, optional and nullable are two separate things. I am not sure if we need optional in our case. Here's my understanding of two.

- `nullable`: I want to explicitly override the existing value by assign `null` to it.
- `optional`: If a new value is not provided, keep the old value as it is. And do not force me to provide a new value.

Reply:
yes, but in my memories if a property is not auto assigned or marked as optional then mikro orm will force you to provide a value. That is why we generally go with an auto assign null which makes it optional without marking it as optional which is what we could eventually do and if we can't we can mark it as optional with a default to null such that you are not forced to provide at least null. I dont know if I am clear 😂 we can discuss it

Reply:
Yeah, I think we should jump on a call to understand it better :)

Reply:
Would it make sense to only have the `nullable` modifier?